gluten-core/src/main/scala/io/glutenproject/extension/columnar/PullOutPreProject.scala

```diff
@@ -149,6 +155,49 @@ object PullOutPreProject extends Rule[SparkPlan] with PullOutProjectHelper {
         agg.child)
       newAgg.withNewChildren(Seq(preProject))
 
+    case window: WindowExec if needsPreProject(window) =>
+      val expressionMap = new mutable.HashMap[Expression, NamedExpression]()
+      // Handle orderSpec.
+      val newOrderSpec = getNewSortOrder(window.orderSpec, expressionMap)
+
+      // Handle partitionSpec.
+      val newPartitionSpec =
+        window.partitionSpec.map(replaceExpressionWithAttribute(_, expressionMap))
+
+      // Handle windowExpressions.
+      val postWindowExpressions = new ArrayBuffer[NamedExpression]()
+      val newWindowExpressions = window.windowExpression.toIndexedSeq
+        .map {
+          _.transform { case we: WindowExpression => rewriteWindowExpression(we, expressionMap) }
+        }
+        .map {
+          case alias @ Alias(_: WindowExpression, _) =>
+            postWindowExpressions += alias.toAttribute
+            alias
+          case other =>
```

Comment:
It should not happen. The window expression must be `Alias(_: WindowExpression, _)`

Reply:
I seem to have encountered a unit test that does not match `Alias(_: WindowExpression, _)`. I will post the corresponding plan later.

Reply:
@ulysses-you `GlutenSQLWindowFunctionSuite.Gluten - Expression in WindowExpression` has a `Alias(Cast(Divide(WindowExpression, Literal)))` case.
```
Window [cast((avg(UnscaledValue(c_acctbal#5)) windowspecdefinition(c_nationkey#4, c_custkey#3 ASC NULLS FIRST, specifiedwindowframe(RangeFrame, unboundedpreceding$(), currentrow$())) / 100.0) as decimal(11,6)) AS avg(c_acctbal) OVER (PARTITION BY c_nationkey ORDER BY c_custkey ASC NULLS FIRST RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW)#10], [c_nationkey#4], [c_custkey#3 ASC NULLS FIRST]
```
<img width="813" alt="image" src="https://github.com/oap-project/gluten/assets/13622031/8bfe415f-7a57-4b72-ae96-37d3b2916edc">

This situation will occur when the input is decimal.

Reply:
ah I see

Reply:
is it possible that we only need post project ? it seems we only go into the pattern if need pre project.

Reply:
@ulysses-you Window and sort are similar, if the `orderSpec` contains an `Expression`, then both pre-project and post-project are definitely required. For `partitionSpec`, only pre-project is needed; for `windowExpression`, only post-project is required, but currently, they are handled together. If we were to extract the `windowExpression` part and handle it separately in the `PullOutPostProject` rule, it could also work. When both `orderSpec` and `windowExpression` contain Expressions, it might produce two post-projects, but the `CollapseProjectTransformer` rule would merge them together. Do you think the part that handles `windowExpression` should be extracted to the `PullOutPostProject` rule?

Reply:
I'm fine to handle post project at here together. My concern is that, the `needsPreProject` method does not cover the case if the window does not need pre project but need post project. We should add post project if the windowExpression child is not the `Alias(_: WindowExpression, _)`.

Reply:
Ok, I know. I think it is better to move this part to `PullOutPostProject` and add this check in `needsPostProject`, I will fix this later.